web/profiles/custom/yalesites_profile/modules/custom/ys_embed/src/Plugin/EmbedSourceBase.php

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * Base class for EmbedSource plugins.
+ *
+ * This plugin is used as a media source for embeded content. It is useful for
+ * social media and streaming meadia providers as well as content shared in
+ * an iframe. The plugin stores the 'input' (raw embed code added by the user),
+ * validates the code, and renders the code through a inline Drupal template.
+ *
+ * New plugins are discovered through annotations. Several are included in this
+ * modules in the ys_embed\Plugin\EmbedSource namespace.
+ */
+abstract class EmbedSourceBase extends PluginBase implements EmbedSourceInterface, ContainerFactoryPluginInterface {
+
+  /**
+   * The media settings config.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $config;
+
+  /**
+   * A regex to match an embed code to the source plugin.
+   *
+   * The regex should used named groups that are stored in the 'params' field
+   * in the database and used to populated the template.
+   *
+   * @var string
+   */
+  protected static $pattern;
+
+  /**
+   * The name of the Drupal template for this code.
+   *
+   * @var string
+   */
+  protected static $template;
+
+  /**
+   * Instructions for finding the embed code in on third party website.
+   *
+   * @var string
+   */
+  protected static $instructions;
+
+  /**
+   * An example of the embed code.
+   *
+   * @var string
+   */
+  protected static $example;
+
+  /**
+   * Creates a plugin instance.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin ID for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory.
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    ConfigFactoryInterface $config_factory
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->config = $config_factory->get('media.settings');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(
+    ContainerInterface $container,
+    array $configuration,
+    $plugin_id,
+    $plugin_definition
+  ) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('config.factory')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDefaultThumbnailUri(): string {
+    $thumbnail = $this->getPluginDefinition()['thumbnail'];
+    return $this->config->get('icon_base_uri') . '/' . $thumbnail;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function isValid(?string $input): bool {
+    return !empty(preg_match(static::$pattern, $input, $matches));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getParams(string $input): array {
+    preg_match(static::$pattern, $input, $matches);
```

Comment:
When I tried to embed a YouTube or other unsupported embed type, it gave the correct error at the bottom, but it also spit out some errors about an empty regular expression. This line requires a pattern and one isn't being sent, I believe. Screenshot attached.
![Screenshot 2022-11-29 at 4 51 57 PM](https://user-images.githubusercontent.com/107938318/204682679-87bc894c-be87-4f9c-8065-51fd7b613ae2.png)

Reply:
Good catch. I did not have warning displaying so I did not see this. And you are correct. So I set a fallback pattern for the 'broken' handler. `protected static $pattern = '/^$/';`